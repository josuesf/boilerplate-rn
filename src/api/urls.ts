export const URL_PRODUCTS =
  'https://6222994f666291106a29f999.mockapi.io/api/v1/products';
// Este tipo de datos debera ser ingresado en un .env segun a variables de entorno con 'react-native-config'
// En fines de crear la prueba sera ubicado en este file.
