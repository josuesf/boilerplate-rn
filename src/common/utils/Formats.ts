import moment from 'moment';

export const FormatDate = (date: Date) => {
  return moment(date).format('D [de] MMMM, YYYY');
};
