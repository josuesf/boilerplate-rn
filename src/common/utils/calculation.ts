import {IProduct} from '../models/Products';

export const getTotalPoints = (data: IProduct[]) =>
  data?.reduce(
    (total, p) =>
      p.is_redemption === false ? total + p.points : total - p.points,
    0,
  );
