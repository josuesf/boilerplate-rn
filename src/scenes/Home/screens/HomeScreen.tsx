import * as React from 'react';
import {Box, Card, Container, Text} from '@components';
import {Header} from '../components/Header';
import {ListPoints} from '../components/ListPoints';
import {useHomeScreen} from '../hooks/useHomeScreen';
import {DropdownSelect} from '../components/DropdownSelect';

export function HomeScreen() {
  const {t, totalPoints, isFetching, data, setFilter} = useHomeScreen();
  return (
    <Container>
      <Box flex={1} mx="sm" mt="l">
        <Header nameUser="Josue Farfan" />
        <Box flex={1}>
          <DropdownSelect setFilter={setFilter} />
          <Text variant="h4">{t('home.yourPoints')}</Text>
          <Card variant="secondary" my="sm">
            <Text color="white" fontWeight="800">
              {t('months.december')}
            </Text>
            <Text mt="s" mb="l" textAlign="center" variant="h1">
              {`${totalPoints < 0 ? '0.00' : totalPoints.toFixed(2)} pts`}
            </Text>
          </Card>
          <Text variant="h4">{t('home.yourMovements')}</Text>
        </Box>
        <ListPoints isLoading={isFetching} data={data || []} />
      </Box>
    </Container>
  );
}
