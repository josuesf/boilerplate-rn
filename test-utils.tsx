import React from 'react';
import {render as rntlRender} from '@testing-library/react-native';
import {ThemeProvider} from './@components';
import moment from 'moment';
import 'moment/locale/es-mx';
import {Provider} from 'react-redux';
import {store} from 'src/redux/store';

function render(ui: React.ReactElement) {
  moment.locale('es-mx');
  return rntlRender(ui, {
    wrapper: ({children}) => (
      <ThemeProvider>
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    ),
  });
}

export * from '@testing-library/react-native';

export {render};
