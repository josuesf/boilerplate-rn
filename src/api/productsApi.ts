import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
import {IProduct} from 'src/common/models/Products';
import {URL_PRODUCTS} from './urls';

export const FILTERS = {
  earned: '?is_redemption=false',
  redemption: '?is_redemption=true',
};

export const productsApi = createApi({
  reducerPath: 'productsApi',
  baseQuery: fetchBaseQuery({
    baseUrl: URL_PRODUCTS,
  }),
  endpoints: builder => ({
    getProducts: builder.query<Array<IProduct>, string>({
      query: filter => `${filter}`,
    }),
  }),
});
export const {useGetProductsQuery} = productsApi;
