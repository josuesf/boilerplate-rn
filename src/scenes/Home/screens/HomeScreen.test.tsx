import React from 'react';
import {HomeScreen} from './HomeScreen';
import {render, screen, act, fireEvent} from 'test-utils';
import dummyProducts from '__tests__/dummyProducts.json';
import {rest} from 'msw';
import {URL_PRODUCTS} from 'src/api/urls';
import {setupServer} from 'msw/node';
import {FormatDate} from 'src/common/utils/formats';

export const handlers = [
  rest.get(URL_PRODUCTS, (_req, res, ctx) => {
    return res(ctx.json(dummyProducts));
  }),
  rest.get(URL_PRODUCTS + '?is_redemption=false', (_req, res, ctx) => {
    return res(ctx.json(dummyProducts.filter(e => e.is_redemption === false)));
  }),
];

export const server = setupServer(...handlers);
describe('HomeScreen', function () {
  // Enable the API mocking before tests.
  beforeAll(() => {
    server.listen();
  });

  // Reset any runtime request handlers we may add during the tests.
  afterEach(() => {
    server.resetHandlers();
    // store.dispatch(apiSlice.util.resetApiState());
  });

  // Disable the API mocking after the tests finished.
  afterAll(() => server.close());

  test('Product find should display data product', async () => {
    const {queryByText, queryByTestId} = render(<HomeScreen />);
    // Show loading data
    expect(queryByTestId('loading')).toBeVisible();
    await act(() => {});
    const rgxName = new RegExp(dummyProducts[0].product, 'i');
    const rgxDate = new RegExp('9 de diciembre, 2022', 'i');
    expect(queryByText(rgxName)).toHaveTextContent(dummyProducts[0].product);
    expect(queryByText(rgxDate)).toHaveTextContent(
      FormatDate(new Date(dummyProducts[0].createdAt)),
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
  test('Show button all when filter is empty and the right Total Points', async () => {
    render(<HomeScreen />);
    const btnEarned = screen.getByTestId('btn-earned');
    expect(btnEarned).toBeVisible();
    expect(screen.getByTestId('btn-redemption')).toBeVisible();
    fireEvent.press(btnEarned);
    await act(() => {});
    expect(screen.getByTestId('btn-all')).toBeVisible();
    //...
  });
});
