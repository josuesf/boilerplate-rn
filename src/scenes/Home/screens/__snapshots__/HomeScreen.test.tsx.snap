// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Product find should display data product 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "marginHorizontal": 39,
          "marginTop": 47,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 39,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#020202",
              "fontFamily": "Avenir",
              "fontSize": 39,
              "fontWeight": "800",
              "lineHeight": 53,
            },
          ]
        }
      >
        home.header.title
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#020202",
              "fontFamily": "Avenir",
              "fontSize": 31,
              "lineHeight": 39,
            },
          ]
        }
        testID="name-user"
      >
        Josue Farfan
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#9B9898",
              "fontFamily": "Avenir",
              "fontSize": 27,
              "fontWeight": "900",
              "lineHeight": 37,
            },
          ]
        }
      >
        home.yourPoints
      </Text>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#334FFA",
              "borderRadius": 39,
              "marginVertical": 39,
              "padding": 39,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Avenir",
                "fontSize": 31,
                "fontWeight": "800",
                "lineHeight": 39,
              },
            ]
          }
        >
          months.december
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Avenir",
                "fontSize": 63,
                "fontWeight": "900",
                "lineHeight": 86,
                "marginBottom": 47,
                "marginTop": 20,
                "textAlign": "center",
              },
            ]
          }
        >
          50.00 pts
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#9B9898",
              "fontFamily": "Avenir",
              "fontSize": 27,
              "fontWeight": "900",
              "lineHeight": 37,
            },
          ]
        }
      >
        home.yourMovements
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 20,
            "flex": 1.2,
            "marginVertical": 31,
            "paddingHorizontal": 20,
            "paddingVertical": 20,
          },
        ]
      }
    >
      <RCTScrollView
        data={
          Array [
            Object {
              "createdAt": "2022-12-09T10:20:00.909Z",
              "id": "3",
              "image": "https://loremflickr.com/640/480/technics",
              "is_redemption": false,
              "points": 200,
              "product": "Fantastic Granite Bacon",
            },
            Object {
              "createdAt": "2022-12-10T10:20:00.909Z",
              "id": "4",
              "image": "https://loremflickr.com/640/480/technics",
              "is_redemption": true,
              "points": 100,
              "product": "My product 2",
            },
            Object {
              "createdAt": "2022-12-20T10:20:00.909Z",
              "id": "5",
              "image": "https://loremflickr.com/640/480/technics",
              "is_redemption": true,
              "points": 50,
              "product": "Mu pRiaaj 20",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        viewabilityConfigCallbackPairs={Array []}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "marginBottom": 20,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#9B9898",
                        "borderRadius": 20,
                        "marginRight": 20,
                      },
                    ]
                  }
                >
                  <Image
                    resizeMethod="auto"
                    resizeMode="cover"
                    source={
                      Object {
                        "uri": "https://loremflickr.com/640/480/technics",
                      }
                    }
                    style={
                      Object {
                        "borderRadius": 20,
                        "height": 108,
                        "width": 108,
                      }
                    }
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0B0B0B",
                          "fontFamily": "Avenir",
                          "fontSize": 27,
                          "fontWeight": "800",
                          "lineHeight": 37,
                        },
                      ]
                    }
                    testID="item-name"
                  >
                    Fantastic Granite Bacon
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0B0B0B",
                          "fontFamily": "Avenir",
                          "fontSize": 24,
                          "fontWeight": "400",
                          "lineHeight": 31,
                        },
                      ]
                    }
                    testID="item-date"
                  >
                    9 de diciembre, 2022
                  </Text>
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#0B0B0B",
                        "fontFamily": "Avenir",
                        "fontSize": 27,
                        "fontWeight": "800",
                        "lineHeight": 37,
                      },
                    ]
                  }
                  testID="item-points"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0ECD9D",
                          "fontFamily": "Avenir",
                          "fontSize": 31,
                          "fontWeight": "800",
                          "lineHeight": 39,
                        },
                      ]
                    }
                  >
                    + 
                  </Text>
                  200
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#020202",
                        "fontFamily": "Avenir",
                        "fontSize": 31,
                        "fontWeight": "800",
                        "lineHeight": 39,
                        "marginLeft": 20,
                      },
                    ]
                  }
                >
                  &gt;
                </Text>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "marginBottom": 20,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#9B9898",
                        "borderRadius": 20,
                        "marginRight": 20,
                      },
                    ]
                  }
                >
                  <Image
                    resizeMethod="auto"
                    resizeMode="cover"
                    source={
                      Object {
                        "uri": "https://loremflickr.com/640/480/technics",
                      }
                    }
                    style={
                      Object {
                        "borderRadius": 20,
                        "height": 108,
                        "width": 108,
                      }
                    }
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0B0B0B",
                          "fontFamily": "Avenir",
                          "fontSize": 27,
                          "fontWeight": "800",
                          "lineHeight": 37,
                        },
                      ]
                    }
                    testID="item-name"
                  >
                    My product 2
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0B0B0B",
                          "fontFamily": "Avenir",
                          "fontSize": 24,
                          "fontWeight": "400",
                          "lineHeight": 31,
                        },
                      ]
                    }
                    testID="item-date"
                  >
                    10 de diciembre, 2022
                  </Text>
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#0B0B0B",
                        "fontFamily": "Avenir",
                        "fontSize": 27,
                        "fontWeight": "800",
                        "lineHeight": 37,
                      },
                    ]
                  }
                  testID="item-points"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#FF0000",
                          "fontFamily": "Avenir",
                          "fontSize": 31,
                          "fontWeight": "800",
                          "lineHeight": 39,
                        },
                      ]
                    }
                  >
                    - 
                  </Text>
                  100
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#020202",
                        "fontFamily": "Avenir",
                        "fontSize": 31,
                        "fontWeight": "800",
                        "lineHeight": 39,
                        "marginLeft": 20,
                      },
                    ]
                  }
                >
                  &gt;
                </Text>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "marginBottom": 20,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#9B9898",
                        "borderRadius": 20,
                        "marginRight": 20,
                      },
                    ]
                  }
                >
                  <Image
                    resizeMethod="auto"
                    resizeMode="cover"
                    source={
                      Object {
                        "uri": "https://loremflickr.com/640/480/technics",
                      }
                    }
                    style={
                      Object {
                        "borderRadius": 20,
                        "height": 108,
                        "width": 108,
                      }
                    }
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0B0B0B",
                          "fontFamily": "Avenir",
                          "fontSize": 27,
                          "fontWeight": "800",
                          "lineHeight": 37,
                        },
                      ]
                    }
                    testID="item-name"
                  >
                    Mu pRiaaj 20
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#0B0B0B",
                          "fontFamily": "Avenir",
                          "fontSize": 24,
                          "fontWeight": "400",
                          "lineHeight": 31,
                        },
                      ]
                    }
                    testID="item-date"
                  >
                    20 de diciembre, 2022
                  </Text>
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#0B0B0B",
                        "fontFamily": "Avenir",
                        "fontSize": 27,
                        "fontWeight": "800",
                        "lineHeight": 37,
                      },
                    ]
                  }
                  testID="item-points"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#FF0000",
                          "fontFamily": "Avenir",
                          "fontSize": 31,
                          "fontWeight": "800",
                          "lineHeight": 39,
                        },
                      ]
                    }
                  >
                    - 
                  </Text>
                  50
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#020202",
                        "fontFamily": "Avenir",
                        "fontSize": 31,
                        "fontWeight": "800",
                        "lineHeight": 39,
                        "marginLeft": 20,
                      },
                    ]
                  }
                >
                  &gt;
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        Array [
          Object {
            "alignContent": "space-between",
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#334FFA",
            "borderRadius": 20,
            "flex": 1,
            "height": 98,
            "justifyContent": "center",
            "marginRight": 20,
            "opacity": 1,
          }
        }
        testID="btn-earned"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Avenir",
                "fontSize": 31,
                "fontWeight": "900",
                "lineHeight": 39,
              },
            ]
          }
        >
          home.btnEarned
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#334FFA",
            "borderRadius": 20,
            "flex": 1,
            "height": 98,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID="btn-redemption"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Avenir",
                "fontSize": 31,
                "fontWeight": "900",
                "lineHeight": 39,
              },
            ]
          }
        >
          home.btnRedemption
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
